src/worker/activities/local_activities.rs

```diff
@@ -0,0 +1,170 @@
+use crate::task_token::TaskToken;
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Formatter},
+};
+use temporal_sdk_core_protos::coresdk::{
+    activity_task::{activity_task, ActivityTask, Start},
+    common::WorkflowExecution,
+    workflow_commands::ScheduleActivity,
+};
+use tokio::sync::{
+    mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender},
+    Mutex, Semaphore,
+};
+
+pub(crate) struct InFlightLocalActInfo {
+    pub seq: u32,
+    pub workflow_execution: WorkflowExecution,
+}
+
+pub(crate) struct NewLocalAct {
+    pub schedule_cmd: ScheduleActivity,
+    pub workflow_type: String,
+    pub workflow_exec_info: WorkflowExecution,
+}
+
+impl Debug for NewLocalAct {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "LocalActivity({}, {})",
+            self.schedule_cmd.seq, self.schedule_cmd.activity_type
+        )
+    }
+}
+
+pub(crate) struct LocalActivityManager {
+    /// Constrains number of currently executing local activities
+    semaphore: Semaphore,
+    /// Sink for new activity execution requests
+    act_req_tx: UnboundedSender<NewLocalAct>,
+
+    dat: Mutex<LAMData>,
+}
+
+struct LAMData {
+    /// Activities that need to be executed by lang
+    act_req_rx: UnboundedReceiver<NewLocalAct>,
+    /// Activities that have been issued to lang but not yet completed
+    outstanding_activity_tasks: HashMap<TaskToken, InFlightLocalActInfo>,
+    next_tt_num: u32,
+}
+
+impl LocalActivityManager {
+    pub(crate) fn new(max_concurrent: usize) -> Self {
+        let (act_req_tx, act_req_rx) = unbounded_channel();
+        Self {
+            semaphore: Semaphore::new(max_concurrent),
+            act_req_tx,
+            dat: Mutex::new(LAMData {
+                act_req_rx,
+                outstanding_activity_tasks: Default::default(),
+                next_tt_num: 0,
+            }),
+        }
+    }
+
+    pub(crate) fn enqueue(&self, acts: impl IntoIterator<Item = NewLocalAct> + Debug) {
+        debug!("Queuing local activities: {:?}", &acts);
+        for act in acts {
+            self.act_req_tx
+                .send(act)
+                .expect("Receive half of LA request channel cannot be dropped");
+        }
+    }
+
+    pub(crate) async fn next_pending(&self) -> Option<ActivityTask> {
+        // Wait for a permit to take a task
+        let permit = self.semaphore.acquire().await.expect("is never closed");
```

Comment:
How does this call ever terminate on a forever-empty channel?

Reply:
It doesn't, and doesn't need to. The caller of this function is selecting between either this or a real activity.

In the shutdown case, the channel will produce a `None` value (but even that doesn't really matter, because the shutdown is also selected upon higher up)

Reply:
Ah, derp, was thinking this was acquisition of work, not acquisition of permit. Nevermind.